Document user files of nix (#8141)
NixOS/nix: doc/manual/src/command-ref/files/user-profiles.md

> # User profiles

Reviewer: Please incorporate my fixes to `nix profile --help` I think you removed by accident. The user of XDG dirs for the profiles directory is unconditional.

Author: Are there any other changes apart from the ones you've mentioned?

Reviewer: I mean #8078, the changes from that PR are gone.

Reviewer: You might also double check #8114 but I think those are OK.

Author: Thanks. I have incorporated them back.

A directory that contains links to user profiles managed by [`nix-env`] and [`nix profile`].

> - `$NIX_STATE_DIR/profiles/per-user/$USER`

Reviewer: Still wrong

> - `${XDG_STATE_DIR-$HOME/.local/state}/nix/profiles` if [`use-xdg-base-directories`] is set to `true`

Reviewer: I'll let you and @fricklerhandwerk be the judge over whether `${XDG_STATE_DIR-$HOME/.local/state}` or plain `$XDG_STATE_DIR` (and maybe a separate XDG reference page) is better, but regardless I think it would be good to be consistent.

> A profile is a directory of symlinks to files in the Nix Store.

Reviewer 2: ```suggestion
A profile is a directory of symlinks to files in the Nix store.
```

## Profile compatibility

> **Warning**
>
> `nix profile` is experimental.

> > Once you have used `nix profile` you can no longer use `nix-env` without first deleting `/nix/var/nix/profiles/per-user/$USER/profile`

Reviewer: still wrong

Once you installed a package with `nix profile`, you get the following
error message when using `nix-env`:

```console
$ nix-env -f '<nixpkgs>' -iA 'hello'
error: --- Error ----------------------------------------------------------------------------------------------------------------- nix-env

> profile '/nix/var/nix/profiles/per-user/<user>/profile' is incompatible with 'nix-env'; please use 'nix profile' instead

Reviewer: still wrong

total 20
lrwxrwxrwx 5 root root 79 Jan  1  1970 chromium -> /nix/store/ijm5k0zqisvkdwjkc77mb9qzb35xfi4m-chromium-86.0.4240.111/bin/chromium
lrwxrwxrwx 7 root root 87 Jan  1  1970 spotify -> /nix/store/w9182874m1bl56smps3m5zjj36jhp3rn-spotify-1.1.26.501.gbe11e53b-15/bin/spotify
lrwxrwxrwx 3 root root 79 Jan  1  1970 zoom-us -> /nix/store/wbhg2ga8f3h87s9h5k0slxk0m81m4cxl-zoom-us-5.3.469451.0927/bin/zoom-us

/home/eelco/.local/state/nix/profiles/profile-7-link/share/applications:
total 12
lrwxrwxrwx 4 root root 120 Jan  1  1970 chromium-browser.desktop -> /nix/store/4cf803y4vzfm3gyk3vzhzb2327v0kl8a-chromium-unwrapped-86.0.4240.111/share/applications/chromium-browser.desktop
lrwxrwxrwx 7 root root 110 Jan  1  1970 spotify.desktop -> /nix/store/w9182874m1bl56smps3m5zjj36jhp3rn-spotify-1.1.26.501.gbe11e53b-15/share/applications/spotify.desktop
lrwxrwxrwx 3 root root 107 Jan  1  1970 us.zoom.Zoom.desktop -> /nix/store/wbhg2ga8f3h87s9h5k0slxk0m81m4cxl-zoom-us-5.3.469451.0927/share/applications/us.zoom.Zoom.desktop

…
```

## Manifest

The manifest file records the provenance of the packages that are installed in this version of the profile.

> It is different between `nix-env` and `nix profile`.

Reviewer 2: ```suggestion
It is different between [`nix-env`] and [`nix profile`].
```

```

Each element in this list corresponds to an installed package.
It incorporates some attributes of the original package derivation, including `meta`, `name`, `out`, `outPath`, `outputs`, `system`.
This can be used for querying and updating the package.

> ### `manifest.json` (used by [`nix profile`])

Reviewer 3: I'm not convinced that moving this from the `nix profile` manpage is an improvement. To me, that's the logical place for this, and it has the advantage of being accessible via `nix profile --help`.

Author: I wanted to keep it a part of `nix profile --help`, but unfortunately due to the way `--help` pages are generated it's not possible to include other files in them.

has the following fields:

* `originalUrl`: The [flake reference](@docroot@/command-ref/new-cli/nix3-flake.md) specified by
  the user at the time of installation (e.g. `nixpkgs`). This is also
  the flake reference that will be used by `nix profile upgrade`.

* `uri`: The locked flake reference to which `originalUrl` resolved.

* `attrPath`: The flake output attribute that provided this
  package. Note that this is not necessarily the attribute that the
  user specified, but the one resulting from applying the default
  attribute paths and prefixes; for instance, `hello` might resolve to
  `packages.x86_64-linux.hello` and the empty string to
  `packages.x86_64-linux.default`.

* `storePath`: The paths in the Nix store containing the package.

* `active`: Whether the profile contains symlinks to the files of this
  package. If set to false, the package is kept in the Nix store, but
  is not "visible" in the profile's symlink tree.

## User profile link

A symbolic link to the user's current profile. 

- `~/.nix-profile`
- `$XDG_STATE_HOME/nix/profile` if [`use-xdg-base-directories`] is set to `true`

This symlink points to:

> - `~/.nix-profile -> prefix/var/nix/profiles/per-user/$USER/profile`

Reviewer: still wrong

- `$XDG_STATE_HOME/nix/profiles` for regular users
- `$NIX_STATE_DIR/profiles/per-user/root` if the user is `root`

A profile is a directory of symlinks to files in the Nix store.

## Profile compatibility

> **Warning**
>
> `nix profile` is experimental.
> Once you have used `nix profile` you can no longer use `nix-env` without first deleting `$XDG_STATE_HOME/nix/profiles/profile`

Once you installed a package with `nix profile`, you get the following error message when using `nix-env`:

```console
$ nix-env -f '<nixpkgs>' -iA 'hello'
error: --- Error ----------------------------------------------------------------------------------------------------------------- nix-env
profile '/home/alice/.local/state/nix/profiles/profile' is incompatible with 'nix-env'; please use 'nix profile' instead
```

To migrate back to `nix-env` you can delete your current profile:

> **Warning**
>
> This will delete packages that have been installed before, so you may want to back up this information before running the command.

```console
 $ rm -rf "${XDG_STATE_HOME-$HOME/.local/state}/nix/profiles/profile"
```

## Filesystem layout

> Profiles are versioned as follows. When using profile *path*, *path*
> is a symlink to *path*`-`*N*, where *N* is the current *version* of
> the profile. In turn, *path*`-`*N* is a symlink to a path in the Nix
> store. For example:

Reviewer 2: ```suggestion
Profiles are versioned as follows. When using profile a named *path*, *path* is a symlink to *path*`-`*N*`-link`, where *N* is the current *version* of the profile.
In turn, *path*`-`*N*`-link` is a symlink to a path in the Nix store.
For example:
```

# User profiles

A directory that contains links to user profiles managed by [`nix-env`] and [`nix profile`].

> - `$NIX_STATE_DIR/profiles/per-user/root` if the user is `root`
> - `$XDG_STATE_HOME/nix/profiles` otherwise

Reviewer 2: ```suggestion
- `$XDG_STATE_HOME/nix/profiles` for regular users
- `$NIX_STATE_DIR/profiles/per-user/root` if the user is `root`
```

A profile is a directory of symlinks to files in the Nix store.

## Profile compatibility

> **Warning**
>
> `nix profile` is experimental.
> Once you have used `nix profile` you can no longer use `nix-env` without first deleting `$XDG_STATE_HOME/nix/profiles/profile`

> Once you installed a package with `nix profile`, you get the following
> error message when using `nix-env`:

Reviewer 2: ```suggestion
Once you installed a package with `nix profile`, you get the following error message when using `nix-env`:
```
